const axios = require('axios');
const https = require('https');

const makeRequestWithRetry = async (url, method, data = {}, maxRetries = 7) => {
    let retries = 0;
    const httpsAgent = new https.Agent({ rejectUnauthorized: false }); // Ignore self-signed certificates

    while (retries < maxRetries) {
        try {
            const response = await axios({ url, method, data, httpsAgent });
            return response.data;
        } catch (error) {
            if (error.code === 'ECONNRESET' || error.response?.status === 500) {
                console.log(`Retry attempt ${retries + 1} due to ${error.code || '500 Internal Server Error'}`);
                retries++;
                continue;
            }
            throw error;
        }
    }
    throw new Error(`Failed to complete request after ${maxRetries} retries.`);
};

module.exports = { makeRequestWithRetry };
